R/BackendHdf5-class.R

```diff
@@ -0,0 +1,225 @@
+                " (current: ", chk[!m], ", stored: ", checksums[!m], ")",
+                collapse="\n"
+            )
+        ))
+    }
+    NULL
+}
+
+.valid.BackendHdf5.h5files <- function(src, h5files) {
+    if (length(src) != length(h5files))
+        return("Different number of hdf5 and original source files.")
+    if (anyDuplicated(h5files))
+        return("Hdf5 files are not unique.")
+    if (!all(file.exists(h5files))) {
+        return(paste0(
+            "File(s) ",
+            paste0(h5files[!file.exists(h5files)], collapse=", "),
+            " not found."
+        ))
+    }
+    NULL
+}
+
+setValidity("BackendHdf5", function(object) {
+    msg <- c(
+        .valid.BackendHdf5.h5files(object@files, object@h5files),
+        .valid.BackendHdf5.checksums(object@checksums, object@h5files)
+    )
+    if (is.null(msg)) { TRUE } else { msg }
+})
+
+#' Initialize the Hdf5 backend, i.e. create the hdf5 files and add their names
+#' to the @hdf5file slot.
+#'
+#' @param object `BackendHdf5`
+#'
+#' @param files `character` with the original file names.
+#'
+#' @param spectraData `DataFrame` with the spectrum metadata.
+#'
+#' @param path `character(1)` defining the path where the hdf5 files should be
+#'     stored.
+#'
+#' @return `BackendHdf5` object.
+#'
+#' @author Johannes Rainer, Sebastian Gibb
+#'
+#' @md
+#'
+#' @noRd
+setMethod(
+    "backendInitialize",
+    "BackendHdf5",
+    function(object, files, spectraData, path=".", ...) {
+    path <- normalizePath(path, mustWork=FALSE)
+    dir.create(path, showWarnings=FALSE, recursive=TRUE)
+
+    files <- normalizePath(files)
+    object@h5files <- file.path(path, paste0(.vdigest(files), ".h5"))
+
+    if (any(file.exists(object@h5files))) {
+        stop(
+            "File(s) ",
+            paste0(object@h5files[file.exists(object@h5files)], collapse=", "),
+            " already exist(s). Please choose a different 'path'."
+        )
+    }
+
+    for (i in seq(along=object@h5files)) {
+        h5 <- H5Fcreate(object@h5files[i])
+        h5createGroup(h5, "spectra")
+        H5Fclose(h5)
+    }
+
+    object@checksums <- .vdigest(object@h5files, file=TRUE)
+
+    callNextMethod()
+})
+
+#' Import the data from the raw MS files and store them to the hdf5 files.
+#'
+#' @inheritParams backendInitialize
+#'
+#' @return `BackendHdf5`
+#'
+#' @author Johannes Rainer, Sebastian Gibb
+#'
+#' @md
+#'
+#' @noRd
+setMethod(
+    "backendImportData",
+    "BackendHdf5",
+    function(object, spectraData, ..., BPPARAM=bpparam()) {
+
+    bpmapply(object@files, object@h5files, FUN=.mzr2hdf5, BPPARAM=BPPARAM)
+    object@checksums <- .vdigest(object@h5files, file=TRUE)
+    validObject(object)
+    object
+})
+
+setMethod(
+    "backendReadSpectra",
+    "BackendHdf5",
+    function(object, spectraData, ...) {
+
+    uFileIdx <- unique(spectraData$fileIdx)
+    h5files <- object@h5files[uFileIdx]
+
+    changed <- .valid.BackendHdf5.checksums(object@checksums[uFileIdx], h5files)
```

Comment:
This is an important test. My only question is if it wouldn't be better to write the file's md5 sum into the hdf5 file (as entry `/md5/md5` as I have in my `BackendHdf5` implementation). How fast is creating the md5 sum of a hdf5 file in contrast to reading it from the file?
 
It's all about performance here. Would be sad if we have a fast backend, but slow it down because of slow data integrity tests.

Reply:
I did a little comparison of md5 sums:
```r
sciex_h5 <- readMSnExperiment(sf, backend = BackendHdf5())
sps <- spectra(filterFile(sciex_mzr, 1))

fl1 <- sciex_h5@backend@h5files[1]
h5_fid <- .Call("_H5Fopen", sciex_h5@backend@h5files[2], 0L, PACKAGE = "rhdf5")

## If we read from the h5 file.
.read_checksum <- function(x) {
    fid <-.Call("_H5Fopen", x, 0L, PACKAGE = "rhdf5")
    on.exit(invisible(.Call("_H5Fclose", fid, PACKAGE = "rhdf5")))
    MSnbase:::.h5_read_bare(fid, "/md5/md5")
}

library(microbenchmark)
microbenchmark(digest::digest(sps),
                   digest::digest(sciex_h5@backend@h5files[1], file = TRUE),
                   .read_checksum(fl),
                   MSnbase:::.h5_read_bare(h5_fid, "/md5/md5"),
                   times = 5)
Unit: microseconds
                                                     expr        min         lq
                                      digest::digest(sps) 174483.582 181643.647
 digest::digest(sciex_h5@backend@h5files[1], file = TRUE)  43428.032  49818.164
                                       .read_checksum(fl)    271.155    350.599
              MSnbase:::.h5_read_bare(h5_fid, "/md5/md5")     45.878     51.455
        mean     median         uq        max neval cld
 194672.5498 193776.073 203786.325 219673.122     5   c
  54820.2748  51157.394  63760.690  65937.094     5  b 
    597.0336    512.348    664.490   1186.576     5 a  
    139.3908     76.466    166.528    356.627     5 a  
```

Summarizing, my way to calculate the checksum on the spectra is pretty bad, calculating checksums on the hdf5 file is faster - but reading the pre-calculated checksum from the hdf5 file beats everything. So I think we should go for that option - ideally on an already opened file.

Reply:
I called my slot `checksums` because I thought we don't want to use `md5` (I think collisions aren't a real problem for us). What about `xxhash32` or `crc32`? At least on my laptop they are 2-3x faster than `md5`.

The problem is that the hash sum calculation on the objects requires all objects in memory (even if you just modified one spectrum). 
There is `utils::fileSnapshot` and `utils::changedFile` that uses `file.info` (ctime, atime, mtime, size) to check whether a file changed (optionally md5 is supported).
We could also use the {a,c,m}time and size and just calculate the checksum if one of the times changed (if the size changed we already know it is different). That should be even faster without the need to store the checksum in the h5 file or to load all the objects into memory.